File: Leetcode/C++/416. Partition Equal Subset Sum.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int sum = 0;
        for(auto a : nums) sum += a;
        
        if(sum%2) return false;
        
        return subset(nums, sum/2);
    }
    
    bool subset(vector<int>& nums, int t) {
        int n = nums.size();
        sort(nums.begin(), nums.end());
        vector<vector<bool>> dp(n+1, vector<bool>(t+1));
        
        for(auto& arr : dp) arr[0] = true;
        
        for(int i=1; i<=n; i++) {
            for(int j=1; j<=t; j++) {
                if(j < nums[i-1]) dp[i][j] = dp[i-1][j];
                else if(i!=0) dp[i][j] = dp[i-1][j] || dp[i-1][j-nums[i-1]];
            }
        }
        
        return dp[n][t];
    }
};

File: Leetcode/TypeScript/416. Partition Equal Subset Sum.ts
function canPartition(nums: number[]): boolean {
    let sum: number = 0;
    for(let i=0; i<nums.length; i++) sum += nums[i];
    
    if(sum%2) return false;
    else return findSubSet(nums, sum/2);
};

const findSubSet = (nums: number[], t: number) => {
    let dp: boolean[][] = [];
    nums.sort((a,b) => a-b);
    
    for(let i=0; i<=nums.length; i++) {
        dp.push([]);
        for(let j=0; j<=t; j++) dp[i].push(false);
    }
    
    for(let i=0; i<dp.length; i++) dp[i][0] = true;
        
    for(let i=1; i<dp.length; i++) {
        for(let j=1; j<=t; j++) dp[i][j] = dp[i-1][j] || (j-nums[i-1] >=0 && dp[i-1][j-nums[i-1]]);
    }
        
    return dp[dp.length-1][dp[0].length-1];
}
